src/postgraphile/http/subscriptions.ts

```diff
@@ -169,151 +177,308 @@ export async function enhanceHttpServerWithSubscriptions<
+              // no meta because validation errors returned from here will be
+              // served through the error message. it contains just the GraphQLErrors
+              // (there is no result to add the meta to)
+            });
+            if (moreValidationRules.length) {
+              const moreValidationErrors = validate(
+                hookedArgs.schema,
+                hookedArgs.document,
+                moreValidationRules,
+              );
+              if (moreValidationErrors.length) {
+                return moreValidationErrors;
+              }
+            }
+
+            return hookedArgs;
+          },
+          async onError(ctx, msg, errors) {
+            // errors returned from onSubscribe
+            releaseContextForSocketAndOpId(ctx.extra.socket, msg.id);
+            const { req, res } = await reqResFromSocket(ctx.extra.socket);
+            return handleErrors(errors, req, res);
+          },
+          async onNext(ctx, _msg, _args, result) {
+            if (result.errors) {
+              // operation execution errors
+              const { req, res } = await reqResFromSocket(ctx.extra.socket);
+              result.errors = handleErrors(result.errors, req, res);
+              return result;
+            }
+          },
+          onComplete(ctx, msg) {
+            releaseContextForSocketAndOpId(ctx.extra.socket, msg.id);
+          },
+        },
+        v1Wss,
+        /*
+         * Heroku times out after 55s:
+         *   https://devcenter.heroku.com/articles/error-codes#h15-idle-connection
+         *
+         * GraphQL Playground times out after 20s:
+         *   https://github.com/prisma/graphql-playground/blob/fa91e1b6d0488e6b5563d8b472682fe728ee0431/packages/graphql-playground-react/src/state/sessions/fetchingSagas.ts#L81
+         *
+         * Pick a number under these ceilings.
+         */
+        subscriptionServerOptions?.keepAlive, // default is 12 seconds
+      );
+    }
+  }
+
+  // listen for upgrades and delegate requests according to the WS subprotocol
+  websocketServer.on('upgrade', (req: IncomingMessage, socket, head) => {
+    const { pathname = '' } = parseUrl(req) || {};
+    const isGraphqlRoute = pathname === graphqlRoute;
+    if (isGraphqlRoute) {
+      const protocol = req.headers['sec-websocket-protocol'];
+      const protocols = Array.isArray(protocol)
+        ? protocol
+        : protocol?.split(',').map(p => p.trim());
+      if (v0Wss && protocols?.includes('graphql-ws')) {
+        const wss = v0Wss;
+        wss.handleUpgrade(req, socket, head, ws => {
+          wss.emit('connection', ws, req);
+        });
+      } else if (v1Wss) {
+        // v1 will welcome its own subprotocol `graphql-transport-ws`
+        // and gracefully reject invalid ones
+        const wss = v1Wss;
+        wss.handleUpgrade(req, socket, head, ws => {
+          wss.emit('connection', ws, req);
+        });
+      }
```

Comment:
If both protocols are supported then v1 should prevail; also please extract the duplicated logic from the `if` statement (e.g. use a ternary to define `wss`)